### DISCUSSION TITLE
cpp/memory/what_is_memory

### DISCUSSION THREAD

**Discussion by giscus** [imarvinle/CSGuide#30] [Announcements]
# cpp/memory/what_is_memory

编程指北: 计算机系统学习指南, 操作系统, 计算机网络, C++, Java, 算法, 数据结构

https://csguide.cn/cpp/memory/what_is_memory.html

**Comment by cocolight** [comment]
受益匪浅，感谢

**Comment by HertZhX** [comment]
999在内存里的4 byte 是不是应该是00000000 | 00000000 | 01110000 | 01111100

**Reply by liguangcheng** [reply]
不对，应该是00000011|11100111，文章作者写错了，存储999，只需要2个字节就可以，不需要四字节，但是计算机一个整型占用4个字节，只是往前面补充0，就可以扩展成4个字节，应该是 00000000|00000000|00000011|11100111

**Reply by liguangcheng** [reply]
用这个程序处理一下就知道了, 高地址的两个字节全0，低地址放的是231和3
```
    int value = 0x3E7; // 999
    for (size_t i = 0; i < sizeof(int); ++i) {
        std::cout << static_cast<unsigned int>(*((reinterpret_cast<unsigned char *>(&value)) + i)) << std::endl;
    }
```
231
3
0
0

**Comment by Guancheng-Li** [comment]
页面有些链接的域名是localhost, 比如“字节序”

**Reply by imarvinle** [reply]
谢谢，已经fix

**Comment by span-t-s** [comment]
> 根据变量的类型和作用域，内存分为几个区域，如栈（stack）、堆（heap）和全局/静态存储区。

C的内存管理对内存划分为 堆、栈、数据段、BSS段、代码段

> 上面只说明了 int 型的变量如何存储在内存，而 float、char 等类型实际上也是一样的，都需要先转换为补码。

有符号整数才会用补码形式存储；float不是用补码形式存储，而是分为符号位、指数位(偏移值形式)、尾数位。
